File: src/main/java/at/bitfire/ical4android/AndroidTaskList.java
/*
 * Copyright (c) 2013 – 2015 Ricki Hirner (bitfire web engineering).
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 */

package at.bitfire.ical4android;

import android.accounts.Account;
import android.annotation.SuppressLint;
import android.content.ContentResolver;
import android.content.ContentUris;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.DatabaseUtils;
import android.net.Uri;
import android.os.RemoteException;

import org.apache.commons.lang3.ArrayUtils;
import org.dmfs.provider.tasks.TaskContract;
import org.dmfs.provider.tasks.TaskContract.TaskLists;
import org.dmfs.provider.tasks.TaskContract.Tasks;

import java.io.FileNotFoundException;
import java.util.LinkedList;
import java.util.List;

import lombok.Cleanup;
import lombok.Getter;

/**
 * Represents a locally stored task list, containing AndroidTasks (whose data objects are Tasks).
 * Communicates with third-party content providers to store the tasks.
 * Currently, these task providers are supported:
 *    - Mirakel
 *    - OpenTasks
 */
public abstract class AndroidTaskList {
    final protected Account account;
    public final TaskProvider provider;
    final AndroidTaskFactory taskFactory;

    @Getter final private long id;
    @Getter private String syncId, name;
    @Getter private Integer color;
    @Getter private boolean isSynced, isVisible;

    /** Those columns will always be fetched when tasks are queried by {@link #queryTasks(String, String[])}.
     *  Must include Tasks._ID as the first element! */
    protected String[] taskBaseInfoColumns() {
        return new String[] { Tasks._ID };
    }


    protected AndroidTaskList(Account account, TaskProvider provider, AndroidTaskFactory taskFactory, long id) {
        this.account = account;
        this.provider = provider;
        this.taskFactory = taskFactory;
        this.id = id;
    }

	
	/* class methods, constructor */

    /**
     * Acquires a ContentProviderClient for a supported task provider. If multiple providers are
     * available, a pre-defined priority list is taken into account.
     * @return A TaskProvider, or null if task storage is not available/accessible.
     *         Caller is responsible for calling release()!
     */
    public static TaskProvider acquireTaskProvider(ContentResolver resolver) {
        TaskProvider.ProviderName[] byPriority = {
                //TaskProvider.ProviderName.Mirakel,
                TaskProvider.ProviderName.OpenTasks
        };
        for (TaskProvider.ProviderName name : byPriority) {
            TaskProvider provider = TaskProvider.acquire(resolver, name);
            if (provider != null)
                return provider;
        }
        return null;
    }


	@SuppressLint("InlinedApi")
	public static Uri create(Account account, TaskProvider provider, ContentValues info) throws CalendarStorageException {
        info.put(TaskContract.ACCOUNT_NAME, account.name);
        info.put(TaskContract.ACCOUNT_TYPE, account.type);
        info.put(TaskLists.ACCESS_LEVEL, 0);

		Constants.log.info("Creating local task list: " + info.toString());
		try {
			return provider.client.insert(syncAdapterURI(provider.taskListsUri(), account), info);
		} catch (RemoteException e) {
			throw new CalendarStorageException("Couldn't create local task list", e);
		}
	}

    public static AndroidTaskList findByID(Account account, TaskProvider provider, AndroidTaskListFactory factory, long id) throws FileNotFoundException, CalendarStorageException {
        try {
            @Cleanup Cursor cursor = provider.client.query(syncAdapterURI(ContentUris.withAppendedId(provider.taskListsUri(), id), account), null, null, null, null);
            if (cursor != null && cursor.moveToNext()) {
                AndroidTaskList taskList = factory.newInstance(account, provider, id);

                ContentValues values = new ContentValues(cursor.getColumnCount());
                DatabaseUtils.cursorRowToContentValues(cursor, values);
                taskList.populate(values);
                return taskList;
            } else
                throw new FileNotFoundException();
        } catch (RemoteException e) {
            throw new CalendarStorageException("Couldn't query task list by ID", e);
        }
    }

    public static AndroidTaskList[] find(Account account, TaskProvider provider, AndroidTaskListFactory factory, String where, String whereArgs[]) throws CalendarStorageException {
        List<AndroidTaskList> taskLists = new LinkedList<>();
        try {
            @Cleanup Cursor cursor = provider.client.query(syncAdapterURI(provider.taskListsUri(), account), null, null, null, null);
            while (cursor != null && cursor.moveToNext()) {
                ContentValues values = new ContentValues(cursor.getColumnCount());
                DatabaseUtils.cursorRowToContentValues(cursor, values);
                AndroidTaskList taskList = factory.newInstance(account, provider, values.getAsLong(TaskLists._ID));
                taskList.populate(values);
                taskLists.add(taskList);
            }
        } catch (RemoteException e) {
            throw new CalendarStorageException("Couldn't query task list by ID", e);
        }
        return taskLists.toArray(factory.newArray(taskLists.size()));
    }

    protected void populate(ContentValues values) {
        syncId = values.getAsString(TaskLists._SYNC_ID);
        name = values.getAsString(TaskLists.LIST_NAME);
        if (values.containsKey(TaskLists.LIST_COLOR))
            color = values.getAsInteger(TaskLists.LIST_COLOR);
        if (values.containsKey(TaskLists.SYNC_ENABLED))
            isSynced = values.getAsInteger(TaskLists.SYNC_ENABLED) != 0;
        if (values.containsKey(TaskLists.VISIBLE))
            isVisible = values.getAsInteger(TaskLists.VISIBLE) != 0;
    }


    public int update(ContentValues info) throws CalendarStorageException {
        try {
            return provider.client.update(taskListSyncUri(), info, null, null);
        } catch (RemoteException e) {
            throw new CalendarStorageException("Couldn't update local task list", e);
        }
    }

    public int delete() throws CalendarStorageException {
        try {
            return provider.client.delete(taskListSyncUri(), null, null);
        } catch (RemoteException e) {
            throw new CalendarStorageException("Couldn't delete local task list", e);
        }
    }


    protected AndroidTask[] queryTasks(String where, String[] whereArgs) throws CalendarStorageException {
        where = (where == null ? "" : "(" + where + ") AND ") + Tasks.LIST_ID + "=?";
        whereArgs = ArrayUtils.add(whereArgs, String.valueOf(id));

        @Cleanup Cursor cursor = null;
        try {
            cursor = provider.client.query(
                    syncAdapterURI(provider.tasksUri()),
                    taskBaseInfoColumns(),
                    where, whereArgs, null);
        } catch (RemoteException e) {
            throw new CalendarStorageException("Couldn't query calendar events", e);
        }

        List<AndroidTask> tasks = new LinkedList<>();
        while (cursor != null && cursor.moveToNext()) {
            ContentValues baseInfo = new ContentValues(cursor.getColumnCount());
            DatabaseUtils.cursorRowToContentValues(cursor, baseInfo);
            tasks.add(taskFactory.newInstance(this, cursor.getLong(0), baseInfo));
        }
        return tasks.toArray(taskFactory.newArray(tasks.size()));
    }


    public static Uri syncAdapterURI(Uri uri, Account account) {
        return uri.buildUpon()
                .appendQueryParameter(TaskContract.ACCOUNT_NAME, account.name)
                .appendQueryParameter(TaskContract.ACCOUNT_TYPE, account.type)
                .appendQueryParameter(TaskContract.CALLER_IS_SYNCADAPTER, "true")
                .build();
    }

    public Uri syncAdapterURI(Uri uri) {
        return uri.buildUpon()
                .appendQueryParameter(TaskContract.ACCOUNT_NAME, account.name)
                .appendQueryParameter(TaskContract.ACCOUNT_TYPE, account.type)
                .appendQueryParameter(TaskContract.CALLER_IS_SYNCADAPTER, "true")
                .build();
    }

    public Uri taskListSyncUri() {
        return syncAdapterURI(ContentUris.withAppendedId(provider.taskListsUri(), id));
    }

}

File: src/main/java/at/bitfire/ical4android/AndroidTaskList.kt
/***************************************************************************************************
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 **************************************************************************************************/

package at.bitfire.ical4android

import android.accounts.Account
import android.content.ContentUris
import android.content.ContentValues
import android.net.Uri
import at.bitfire.ical4android.util.MiscUtils.CursorHelper.toValues
import at.bitfire.ical4android.util.MiscUtils.UriHelper.asSyncAdapter
import org.dmfs.tasks.contract.TaskContract
import org.dmfs.tasks.contract.TaskContract.Property.Relation
import org.dmfs.tasks.contract.TaskContract.TaskLists
import org.dmfs.tasks.contract.TaskContract.Tasks
import java.io.FileNotFoundException
import java.util.*


/**
 * Represents a locally stored task list, containing AndroidTasks (whose data objects are Tasks).
 * Communicates with third-party content providers to store the tasks.
 * Currently, only the OpenTasks tasks provider (org.dmfs.provider.tasks) is supported.
 */
abstract class AndroidTaskList<out T : AndroidTask>(
    val account: Account,
    val provider: TaskProvider,
    val taskFactory: AndroidTaskFactory<T>,
    val id: Long
) {

    companion object {

        fun create(account: Account, provider: TaskProvider, info: ContentValues): Uri {
            info.put(TaskContract.ACCOUNT_NAME, account.name)
            info.put(TaskContract.ACCOUNT_TYPE, account.type)
            info.put(TaskLists.ACCESS_LEVEL, 0)

            Ical4Android.log.info("Creating local task list: $info")
            return provider.client.insert(provider.taskListsUri().asSyncAdapter(account), info)
                ?: throw CalendarStorageException("Couldn't create task list (empty result from provider)")
        }

        fun <T : AndroidTaskList<AndroidTask>> findByID(
            account: Account,
            provider: TaskProvider,
            factory: AndroidTaskListFactory<T>,
            id: Long
        ): T {
            provider.client.query(
                ContentUris.withAppendedId(provider.taskListsUri(), id).asSyncAdapter(account),
                null,
                null,
                null,
                null
            )?.use { cursor ->
                if (cursor.moveToNext()) {
                    val taskList = factory.newInstance(account, provider, id)
                    taskList.populate(cursor.toValues())
                    return taskList
                }
            }
            throw FileNotFoundException()
        }

        fun <T : AndroidTaskList<AndroidTask>> find(
            account: Account,
            provider: TaskProvider,
            factory: AndroidTaskListFactory<T>,
            where: String?,
            whereArgs: Array<String>?
        ): List<T> {
            val taskLists = LinkedList<T>()
            provider.client.query(
                provider.taskListsUri().asSyncAdapter(account),
                null,
                where,
                whereArgs,
                null
            )?.use { cursor ->
                while (cursor.moveToNext()) {
                    val values = cursor.toValues()
                    val taskList =
                        factory.newInstance(account, provider, values.getAsLong(TaskLists._ID))
                    taskList.populate(values)
                    taskLists += taskList
                }
            }
            return taskLists
        }

    }

    var syncId: String? = null
    var name: String? = null
    var color: Int? = null
    var isSynced = false
    var isVisible = false


    protected fun populate(values: ContentValues) {
        syncId = values.getAsString(TaskLists._SYNC_ID)
        name = values.getAsString(TaskLists.LIST_NAME)
        color = values.getAsInteger(TaskLists.LIST_COLOR)
        values.getAsInteger(TaskLists.SYNC_ENABLED)?.let { isSynced = it != 0 }
        values.getAsInteger(TaskLists.VISIBLE)?.let { isVisible = it != 0 }
    }

    fun update(info: ContentValues) = provider.client.update(taskListSyncUri(), info, null, null)
    fun delete() = provider.client.delete(taskListSyncUri(), null, null)

    /**
     * When tasks are added or updated, they may refer to related tasks by UID ([Relation.RELATED_UID]).
     * However, those related tasks may not be available (for instance, because they have not been
     * synchronized yet), so that the tasks provider can't establish the actual relation (= set
     * [Relation.TASK_ID]) in the database.
     *
     * As soon as such a related task is added, OpenTasks updates the [Relation.RELATED_ID],
     * but it does *not* update [Tasks.PARENT_ID] of the parent task:
     * https://github.com/dmfs/opentasks/issues/877
     *
     * This method shall be called after all tasks have been synchronized. It touches
     *
     *   - all [Relation] rows
     *   - with [Relation.RELATED_ID] (→ related task is already synchronized)
     *   - of tasks without [Tasks.PARENT_ID] (→ only touch relevant rows)
     *
     * so that missing [Tasks.PARENT_ID] fields are updated.
     *
     * @return number of touched [Relation] rows
     */
    fun touchRelations(): Int {
        Ical4Android.log.fine("Touching relations to set parent_id")
        val batchOperation = BatchOperation(provider.client)
        provider.client.query(
            tasksSyncUri(true), null,
            "${Tasks.LIST_ID}=? AND ${Tasks.PARENT_ID} IS NULL AND ${Relation.MIMETYPE}=? AND ${Relation.RELATED_ID} IS NOT NULL",
            arrayOf(id.toString(), Relation.CONTENT_ITEM_TYPE),
            null, null
        )?.use { cursor ->
            while (cursor.moveToNext()) {
                val values = cursor.toValues()
                val id = values.getAsLong(Relation.PROPERTY_ID)
                val propertyContentUri = ContentUris.withAppendedId(tasksPropertiesSyncUri(), id)
                batchOperation.enqueue(
                    BatchOperation.CpoBuilder
                        .newUpdate(propertyContentUri)
                        .withValue(Relation.RELATED_ID, values.getAsLong(Relation.RELATED_ID))
                )
            }
        }
        return batchOperation.commit()
    }


    /**
     * Queries tasks from this task list. Adds a WHERE clause that restricts the
     * query to [Tasks.LIST_ID] = [id].
     *
     * @param _where selection
     * @param _whereArgs arguments for selection
     *
     * @return events from this task list which match the selection
     */
    fun queryTasks(_where: String? = null, _whereArgs: Array<String>? = null): List<T> {
        val where = "(${_where ?: "1"}) AND ${Tasks.LIST_ID}=?"
        val whereArgs = (_whereArgs ?: arrayOf()) + id.toString()

        val tasks = LinkedList<T>()
        provider.client.query(
            tasksSyncUri(),
            null,
            where, whereArgs, null
        )?.use { cursor ->
            while (cursor.moveToNext())
                tasks += taskFactory.fromProvider(this, cursor.toValues())
        }
        return tasks
    }

    fun findById(id: Long) = queryTasks("${Tasks._ID}=?", arrayOf(id.toString())).firstOrNull()
        ?: throw FileNotFoundException()


    fun taskListSyncUri() =
        ContentUris.withAppendedId(provider.taskListsUri(), id).asSyncAdapter(account)

    fun tasksSyncUri(loadProperties: Boolean = false): Uri {
        val uri = provider.tasksUri().asSyncAdapter(account)
        return if (loadProperties)
            uri.buildUpon()
                .appendQueryParameter(TaskContract.LOAD_PROPERTIES, "1")
                .build()
        else
            uri
    }

    fun tasksPropertiesSyncUri() = provider.propertiesUri().asSyncAdapter(account)

}
